Sources/Dip.swift

```diff
@@ -578,3 +579,67 @@ extension DependencyContainer.Tag: Equatable {
   }
 
 }
+
+
+final class DefinitionsContainer {
+  // Backing data structure. Its a map of
+  // [Type.hashValue: [DefinitionKey: _Definition]]
+  // so we have fash lookup of by Type for the definitions that handle it.
+  private var definitionsByType = [Int: [DefinitionKey: _Definition]]()
+  
+  typealias DefinitionKeyValuePair = (DefinitionKey, _Definition)
+  
+  var all: [DefinitionKeyValuePair] {
+    return definitionsByType.reduce(into: [DefinitionKeyValuePair]()) { (acc, arg1) in
+      let (_, value) = arg1
+      acc.append(contentsOf: value.map { ($0, $1)})
+    }
+  }
+  
+  func removeAll() {
+      definitionsByType.removeAll()
+  }
+  
+  func add<T,U>(key: DefinitionKey, definition: Definition<T, U>) {
+    do {
+      let typeKey = ObjectIdentifier(T.self).hashValue
+      var definitionMap = definitionsByType[typeKey] ?? [:]
+      definitionMap[key] = definition
+      definitionsByType[typeKey] = definitionMap
+    }
+    
+    // Definitions automatically store T? so so does our table
```

Comment:
does this mean the above block of code doesn't do anything?

Reply:
They automatically store BOTH `T` and `T?` so that you can resolve optionals:
`let service: Service? = container.resolver() as Service?`